script/watch.js

```diff
@@ -0,0 +1,170 @@
+      groupedFiles.eslint :
+      await glob('+(src|test)/**/*.js?(x)');
+    console.log(`Running Eslint with ${eslintTargets.length} Files\n`);
+    eslintFiles(eslintTargets);
+  }
+
+  if (options.sasslint) {
+    const sasslintTargets = groupedFiles.sasslint.length > 0 ?
+      groupedFiles.sasslint :
+      await glob('./**/*.scss');
+
+    console.log(`Running sassLint with ${sasslintTargets.length} Files\n`);
+    sasslintFiles(sasslintTargets);
+  }
+
+  console.log(chalk.cyan('Watching for changes'));
+}
+
+function groupFilesByTest(files) {
+  return files.reduce((groupedFiles, file) => {
+    if (file.includes('.unit.spec')) {
+      groupedFiles.mocha.push(file);
+    }
+
+    if (!file.includes('src') && !file.includes('test')) {
+      return groupedFiles;
+    }
+
+    switch(path.extname(file)) {
+      case '.js':
+      case '.jsx':
+        groupedFiles.eslint.push(file);
+        break;
+      case '.scss':
+        groupedFiles.sasslint.push(file);
+        break;
+      default:
+    }
+    return groupedFiles;
+
+  }, { eslint: [], mocha: [], sasslint: [] } );
+}
+
+async function getChangedFiles() {
+  const projectRoot = path.resolve();
+  // use gitDiffIndex to limit tests that are run to changed files
+  const gitDiffIndex = await exec('git diff-index --name-only HEAD');
+  return gitDiffIndex.stdout.split('\n').map(file => path.join(projectRoot, file));
+}
+
+async function runMochaTests(files) {
+  // keep errors from polluting the reporter by redirecting the error stream
+  let errorStream;
+  const consoleErr = process.stderr.write;
+  if (!options.showErrors) {
+    errorStream = new Writable({
+      write(chunk, encoding, callback) {
+        callback();
+      }
+    });
+    process.stderr.write = errorStream;
+  }
+
+  const mocha = new Mocha({ignoreLeaks: true});
+  mocha.reporter('nyan');
+  mocha.addFile('./test/helper.js');
+
+  files.forEach(file => mocha.addFile(file));
+  return new Promise ((fulfill, reject) => {
+    mocha.run()
+      .once('end', function() {
+        // reattach error stream
+        process.stderr.write = consoleErr;
+        fulfill();
+      });
+  });
+}
+
+function eslintFiles(files) {
+  const eslintResults = eslint
+    .executeOnFiles(files)
+    .results
+    .filter(result => options.showErrors || result.errorCount > 0);
+  console.log(formatter(eslintResults) || chalk.green(`${files.length} js files lint free`));
+}
+
+function sasslintFiles(files) {
+  const sasslintResults = _.flattenDeep(files.map(sassFile =>
+    sasslinter.lintFiles(`./${sassFile}`, {}, './config/sass-lint.yml')
+  ));
+  console.log(formatter(sasslintResults) || chalk.green(`${files.length} scss files lint free`));
+}
+
+// ensures errors that blow up test watcher make it to console reporter
+process.on('uncaughtException', function(err) {
+  console.error((err && err.stack) ? err.stack : err);
+});
+srcWatcher.on('change', runTests);
+runTests();
```

Comment:
It seems like it's hanging with "Running Mocha with 486 Files" when it first starts up. Should it be running all the tests, or do I need to change a file first?

Reply:
It should run all of the tests when it starts up. It does hang for a few seconds while mocha requires and babelifies all of the files. I can provide more feedback or put the initial run behind a flag- this is a personal preference for me. I like to see if there are any failing tests when I start the watch task.